nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/JmxMetricsResource.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.nifi.web.NiFiServiceFacade;
+import org.apache.nifi.web.api.metrics.jmx.JmxMetricsCollector;
+import org.apache.nifi.web.api.metrics.jmx.JmxMetricsFilter;
+import org.apache.nifi.web.api.metrics.jmx.JmxMetricsResult;
+import org.apache.nifi.web.api.metrics.jmx.JmxMetricsResultConverter;
+import org.apache.nifi.web.api.metrics.jmx.JmxMetricsWriter;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.StreamingOutput;
+import java.util.Collection;
+
+/**
+ * RESTful endpoint for JMX metrics.
+ */
+@Path("/jmx-metrics")
+@Api(
+        value = "/jmx-metrics",
+        description = "Endpoint for accessing the JMX metrics."
+)
+public class JmxMetricsResource extends ApplicationResource {
+    private static final String JMX_METRICS_NIFI_PROPERTY = "nifi.jmx.metrics.blacklisting.filter";
+    private NiFiServiceFacade serviceFacade;
+    private Authorizer authorizer;
+
+    /**
+     * Retrieves the JMX metrics.
+     *
+     * @return A jmxMetricsResult list.
+     */
+    @GET
+    @Consumes(MediaType.WILDCARD)
+    @Produces(MediaType.WILDCARD)
+    @ApiOperation(
+            value = "Gets all allowed JMX metrics",
+            response = StreamingOutput.class,
+            authorizations = {
+                    @Authorization(value = "Read - /flow"),
+                    @Authorization(value = "Read - /system-diagnostics")
+            }
+    )
+    @ApiResponses(
+            value = {
+                    @ApiResponse(code = 400, message = "NiFi was unable to complete the request because it was invalid. The request should not be retried without modification."),
+                    @ApiResponse(code = 401, message = "Client could not be authenticated."),
+                    @ApiResponse(code = 403, message = "Client is not authorized to make this request."),
+                    @ApiResponse(code = 404, message = "The specified resource could not be found."),
+                    @ApiResponse(code = 409, message = "The request was valid but NiFi was not in the appropriate state to process it. Retrying the same request later may be successful.")
+            }
+    )
+    public Response getJmxMetrics(
+            @ApiParam(
+                    value = "Regular Expression Pattern to be applied against the ObjectName")
+            @QueryParam("beanNameFilter") final String beanNameFilter
+
+    ) {
+        authorizeJmxMetrics();
+
+        final String blackListingFilter = getProperties().getProperty(JMX_METRICS_NIFI_PROPERTY);
+        final JmxMetricsResultConverter metricsResultConverter = new JmxMetricsResultConverter();
+        final JmxMetricsCollector jmxMetricsCollector = new JmxMetricsCollector(metricsResultConverter);
+
+        final Collection<JmxMetricsResult> results = jmxMetricsCollector.getBeanMetrics();
```

Comment:
Although the implementation is fairly straightforward, in general there should be some decoupling between REST Resource methods and implementations. Recommend introducing a `JmxMetricsService` interface to encapsulate the collection and filtering, and provide the implementation through property wiring.